Package support policy (#76)
dotnet/maintenance-packages: package-support-policy.md

| System.Threading.Tasks.Dataflow                              | Upgrade          | Inbox in latest                                     | Inbox in latest                                     |
| System.Threading.Tasks.Extensions                            | Remove           | Migrated                                            | Migrated                                            |
| System.Threading.Tasks.Parallel                              | Remove           | .NET Standard 2.0 implementation, not used anywhere | .NET Standard 2.0 implementation, not used anywhere |
| System.Threading.Thread                                      | Remove           | Façade                                              | Façade                                              |
| System.Threading.ThreadPool                                  | Remove           | Façade                                              | Façade                                              |
| System.Threading.Timer                                       | Remove           | Façade                                              | Façade                                              |

> | System.ValueTuple                                            | Remove           | .NET Framework implementation with in-box façade    | .NET Framework implementation with in-box façade    |

Reviewer: System.ValueTuple is needed for .NET Framework 4.6.2

> | System.ValueTuple                                            | Remove           | .NET Framework implementation with in-box façade    | .NET Framework implementation with in-box façade    |

Reviewer: System.ValueTuple is needed for .NET Framework 4.6.2

| System.Xml.ReaderWriter                                      | Remove           | .NET Standard 2.0 implementation, not used anywhere | .NET Standard 2.0 implementation, not used anywhere |
| System.Xml.XDocument                                         | Remove           | .NET Standard 2.0 implementation, not used anywhere | .NET Standard 2.0 implementation, not used anywhere |
| System.Xml.XmlDocument                                       | Remove           | .NET Standard 2.0 implementation, not used anywhere | .NET Standard 2.0 implementation, not used anywhere |
| System.Xml.XmlSerializer                                     | Remove           | .NET Standard 2.0 implementation, not used anywhere | .NET Standard 2.0 implementation, not used anywhere |
| System.Xml.XPath                                             | Remove           | .NET Standard 2.0 implementation, not used anywhere | .NET Standard 2.0 implementation, not used anywhere |
| System.Xml.XPath.XDocument                                   | Remove           | Simple implementation                               | Simple implementation                               |

> | System.Xml.XPath.XmlDocument                                 | Remove           | Migrated                                            | Migrated                                            |

Reviewer 2: @carlossanlop I don't like that we are adding all packages to this document. I think it makes sense documenting the packages that we brought over or intend to bring over but I wouldn't mention any of our actively developed packages like Microsoft.Extensions.*

# Package support policy for maintenance-packages

As it has been described in the [.NET Release policies](https://github.com/dotnet/core/blob/main/release-policies.md) document, .NET library packages provide additional functionality that complements the .NET libraries and application frameworks like ASP.NET Core. The library packages have the same [support model as the .NET versions](https://dotnet.microsoft.com/en-us/platform/support/policy) that they target.

For example, [Microsoft.Extensions.Logging](https://www.nuget.org/packages/Microsoft.Extensions.Logging/#supportedframeworks-body-tab) supports multiple .NET versions via the frameworks it targets.

Our packages support any target framework versions that are currently in-support. You can find out which frameworks are supported by a specific package by visiting its NuGet page and navigating to the frameworks tab.

Additionally, you can find out which target frameworks are currently in-support by visiting the following pages:

- .NET: https://dotnet.microsoft.com/en-us/platform/support/policy

> - .NET Framework: https://dotnet.microsoft.com/en-us/learn/dotnet/what-is-dotnet-framework

Reviewer 3: Should we link to https://learn.microsoft.com/en-us/lifecycle/products/microsoft-net-framework?

Author: I guess that is the document that would make more sense to link to, since it's the one that contains the support policy and also has a link to the lifecycle FAQ.

I originally opted for the [what-is-dotnet-framework](https://dotnet.microsoft.com/en-us/learn/dotnet/what-is-dotnet-framework) doc because it lists the supported versions, the out of support versions, and also a link to the support policy at the end.

I'm not strongly inclined to any option, so if you strongly feel like the support policy doc is the one that needs to be linked directly, let me know and I'll apply the suggestion below.

Or we can do it democratically 😄 . @ViktorHofer what do you think? We need your tie-breaking vote.

```suggestion
- .NET Framework: https://learn.microsoft.com/en-us/lifecycle/products/microsoft-net-framework
```

> - .NET Standard: https://learn.microsoft.com/en-us/dotnet/standard/net-standard

Reviewer 3: .NETStandard isn't a runnable framework and shouldn't be listed here.

Author: ```suggestion
```

On the other hand, the packages that come out from the maintenance-packages repo have a separate, independent release cadence, meaning their servicing releases will come out of this repo and will not be bound by any pre-defined release schedule.

These are packages that originally had their source code hosted in branches that are already out of support, but the packages themselves are still in-support, because they target frameworks that are still supported. Some examples:

- [System.Xml.XPath.XmlDocument](https://www.nuget.org/packages/System.Xml.XPath.XmlDocument) - Its source code used to be hosted in the .NET Core 1.1 branch.
- [System.Buffers](https://www.nuget.org/packages/System.Buffers) - Its source code used to be hosted in the .NET Core 2.1 branch.
- [Microsoft.Bcl.HashCode](https://www.nuget.org/packages/Microsoft.Bcl.HashCode) - Its source code used to be hosted in the.NET Core 3.1 branch.
- [System.Net.WebSockets.WebSocketProtocol](https://www.nuget.org/packages/System.Net.WebSockets.WebSocketProtocol) - Its source code used to be hosted in the .NET 5.0 branch.

## End of support

> A package will be supported for as long as it targets frameworks that are still supported. For example: [System.Numerics.Vectors](https://www.nuget.org/packages/System.Json#supportedframeworks-body-tab). Of all the included target frameworks in the package, .NET Standard 2.0 is still in support at the moment of writing this document. As long as there are frameworks compatible with .NET Standard 2.0, the package will remain in support.

Reviewer 3: NETStandard is not sufficient here, since it's not a runnable framework.  We also don't explicitly "support" NETStandard with servicing because of this - you can't run code there. 

Packages are supported on the frameworks where they can actually run.  `System.Numerics.Vectors` is supported because it provides an implementation on `net462` which is still in support.

Author: > NETStandard with servicing because of this - you can't run code there.

Right, but that's why the last sentence says:

> As long as there are frameworks compatible with .NET Standard 2.0, the package will remain in support.

Would that be confusing to customers? What would you suggest instead?

| [System.Buffers](src/System.Buffers/)                                                   | 2.1                  |
| [System.Data.SqlClient](src/System.Data.SqlClient/)                                     | 3.1                  |
| [System.Json](src/System.Json/)                                                         | 3.1                  |
| [System.Memory](src/System.Memory/)                                                     | 2.1                  |
| [System.Net.WebSockets.WebSocketProtocol](src/System.Net.WebSockets.WebSocketProtocol/) | 5.0                  |
| [System.Numerics.Vectors](src/System.Numerics.Vectors/)                                 | 2.1                  |
| [System.Reflection.DispatchProxy](src/System.Reflection.DispatchProxy/)                 | 3.1                  |
| [System.Runtime.CompilerServices.Unsafe](src/System.Runtime.CompilerServices.Unsafe/)   | 6.0                  |
| [System.Threading.Tasks.Extensions](src/System.Threading.Tasks.Extensions/)             | 2.1                  |
| [System.Xml.XPath.XmlDocument](src/System.Xml.XPath.XmlDocument/)                       | 1.1                  |

> ### Implementation too simple

Reviewer 3: These tables show how you've decided whether or not to include a package in this repo, but that doesn't really answer _customer_ questions they might have about a package.  It's also difficult to navigate all these.

Here are some customer questions I'd hope we could answer with a table.
1. Should I still reference package X on framework Y?  (Maybe we could have this for each minium supported framework and NETStandard)
2. Why isn't there a new version of package X?

Maybe we could have a single table that covers this?

Author: From our chat today, I summarized the explanation:
If we had a single table that listed every package that we ever shipped or that we would expect to be covered by maintenance-packages, for the ones that are still in service we say update to latest, for the others we can tell folks not to reference it anymore, or yes you can reference and it does not need to change.

|-----------------------------------------|----------------------|
| System.IO.FileSystem.AccessControl      | 5.0                  |
| System.IO.Pipes.AccessControl           | 5.0                  |
| System.Reflection.TypeExtensions        | 3.1                  |
| System.Data.Common                      | 1.1                  |
| System.Diagnostics.StackTrace           | 1.1                  |
| System.Globalization.Extensions         | 1.1                  |
| System.Net.Http                         | 1.1                  |
| System.Net.Sockets                      | 1.1                  |
| System.Runtime.Serialization.Primitives | 1.1                  |
| System.Security.Cryptography.Algorithms | 1.1                  |
| System.Xml.XPath.XDocument              | 1.1                  |

### No code

These packages contain no code.

| Package                            | Last shipped release |
|------------------------------------|----------------------|
| Microsoft.NETCore.Platforms.Future | 3.1                  |
| Microsoft.NETCore.Targets          | 5.0                  |
| Microsoft.NETCore.Platforms        | 7.0                  |

> | Microsoft.Windows.Compatibility    | 8.0                  |

Reviewer 3: Any packages still being built can be omitted from these docs. 
```suggestion
```

| Microsoft.Extensions.Logging.Configuration                   | 8.0                  |
| Microsoft.Extensions.Logging.Console                         | 8.0                  |
| Microsoft.Extensions.Logging.Debug                           | 8.0                  |
| Microsoft.Extensions.Logging.EventLog                        | 8.0                  |
| Microsoft.Extensions.Logging.EventSource                     | 8.0                  |
| Microsoft.Extensions.Logging.TraceSource                     | 8.0                  |
| Microsoft.Extensions.Options                                 | 8.0                  |
| Microsoft.Extensions.Options.ConfigurationExtensions         | 8.0                  |
| Microsoft.Extensions.Options.DataAnnotations                 | 8.0                  |
| Microsoft.Extensions.Primitives                              | 8.0                  |
| Microsoft.ILVerification                                     | 8.0                  |
| Microsoft.NET.ILLink                                         | 8.0                  |
| Microsoft.NET.ILLink.Tasks                                   | 8.0                  |

> | Microsoft.NET.Sdk.IL                                         | 8.0                  |
> | Microsoft.NET.WebAssembly.Threading                          | 8.0                  |
> | Microsoft.NETCore.ILAsm                                      | 8.0                  |
> | Microsoft.NETCore.ILDAsm                                     | 8.0                  |
> | Microsoft.Win32.Registry.AccessControl                       | 8.0                  |
> | Microsoft.Win32.SystemEvents                                 | 8.0                  |
> | Microsoft.XmlSerializer.Generator                            | 8.0                  |
> | System.CodeDom                                               | 8.0                  |
> | System.Collections.Immutable                                 | 8.0                  |
> | System.ComponentModel.Composition                            | 8.0                  |
> | System.ComponentModel.Composition.Registration               | 8.0                  |
> | System.Composition                                           | 8.0                  |
> | System.Composition.AttributedModel                           | 8.0                  |
> | System.Composition.Convention                                | 8.0                  |
> | System.Composition.Hosting                                   | 8.0                  |
> | System.Composition.Runtime                                   | 8.0                  |
> | System.Composition.TypedParts                                | 8.0                  |
> | System.Configuration.ConfigurationManager                    | 8.0                  |
> | System.Data.Odbc                                             | 8.0                  |
> | System.Data.OleDb                                            | 8.0                  |
> | System.Diagnostics.DiagnosticSource                          | 8.0                  |
> | System.Diagnostics.EventLog                                  | 8.0                  |
> | System.Diagnostics.PerformanceCounter                        | 8.0                  |
> | System.DirectoryServices                                     | 8.0                  |
> | System.DirectoryServices.AccountManagement                   | 8.0                  |
> | System.DirectoryServices.Protocols                           | 8.0                  |
> | System.Drawing.Common                                        | 8.0                  |
> | System.Formats.Asn1                                          | 8.0                  |
> | System.Formats.Cbor                                          | 8.0                  |
> | System.IO.Hashing                                            | 8.0                  |
> | System.IO.Packaging                                          | 8.0                  |
> | System.IO.Pipelines                                          | 8.0                  |
> | System.IO.Ports                                              | 8.0                  |
> | System.Management                                            | 8.0                  |
> | System.Memory.Data                                           | 8.0                  |
> | System.Net.Http.Json                                         | 8.0                  |
> | System.Net.Http.WinHttpHandler                               | 8.0                  |
> | System.Numerics.Tensors                                      | 8.0                  |
> | System.Reflection.Context                                    | 8.0                  |
> | System.Reflection.Metadata                                   | 8.0                  |
> | System.Reflection.MetadataLoadContext                        | 8.0                  |
> | System.Resources.Extensions                                  | 8.0                  |
> | System.Runtime.Caching                                       | 8.0                  |
> | System.Runtime.Serialization.Schema                          | 8.0                  |
> | System.Security.Cryptography.Cose                            | 8.0                  |
> | System.Security.Cryptography.Pkcs                            | 8.0                  |
> | System.Security.Cryptography.ProtectedData                   | 8.0                  |
> | System.Security.Cryptography.Xml                             | 8.0                  |
> | System.Security.Permissions                                  | 8.0                  |
> | System.ServiceModel.Syndication                              | 8.0                  |
> | System.ServiceProcess.ServiceController                      | 8.0                  |
> | System.Speech                                                | 8.0                  |
> | System.Text.Encoding.CodePages                               | 8.0                  |
> | System.Text.Encodings.Web                                    | 8.0                  |
> | System.Text.Json                                             | 8.0                  |
> | System.Threading.AccessControl                               | 8.0                  |
> | System.Threading.Channels                                    | 8.0                  |
> | System.Threading.RateLimiting                                | 8.0                  |
> | System.Threading.Tasks.Dataflow                              | 8.0                  |
> | System.Windows.Extensions                                    | 8.0                  |

Author: From our conversation @ericstj, I should remove the packages that we build in the latest version.

I will keep those that we stopped building in older versions though (the 3 ones at the end of the table).

```suggestion
```

| System.Security.AccessControl                                | 6.0                  |
| Microsoft.Diagnostics.Tracing.EventSource.Redist             | 6.0                  |
| Microsoft.IO.Redist                                          | 6.0                  |

### Others

These packages are fully unsupported:

| Package                             | Last shipped release |
|-------------------------------------|----------------------|
| System.Diagnostics.Tracing          | 1.1                  |
| System.Reflection.Emit              | 3.1                  |
| System.Reflection.Emit.ILGeneration | 3.1                  |

> | System.Reflection.Emit.Lightweight  | 3.1                  |

Author: Looking at these again:

- For System.Diagnostics.Tracing we've been providing workarounds and don't want to support this package anymore. So I could put it in a separate table indicating this info.

For System.Reflection.Emit* it seems we made them in-box (there used to be a `pkg/` folder in 3.1, and we didn't add the `GeneratePackageOnBuild` in 5.0). Which means I could move them to the `## In-box` section above instead.

Author: Based on the other feedback, this table will change considerably so I'll resolve this.

- [System.Text.Json](https://www.nuget.org/packages/System.Text.Json), which is part of the shared framework, but is also available as a NuGet package.
- [Microsoft.Extensions.Configuration.Abstractions](https://www.nuget.org/packages/Microsoft.Extensions.Configuration.Abstractions), which is exclusively released as a NuGet package.

Note that the source code of those two packages lives in the [dotnet/runtime](https://github.com/dotnet/runtime) repo.

On the other hand, the packages that come out from the dotnet/maintenance-packages repo have a separate, independent release cadence, meaning their servicing releases will come out of this repo and will not be bound by any pre-defined release schedule.

## End of support

A package will be supported for as long as it targets frameworks that are still supported. For example: [System.Numerics.Vectors](https://www.nuget.org/packages/System.Json#supportedframeworks-body-tab) is supported because it provides an implementation on .NET Framework 4.6.2, which is still in support.

## Reporting issues

> If you find an issue in an assembly that lives in this repo, please report it in our central repo: https://github.com/dotnet/runtime/issues/new/choose.

Reviewer 2: Interesting. Why don't we want issues filed in this repo?

Author: Visibility. The team mostly works in runtiime. Only you, Eric and I actively look at this repo.

## Should I consume this package? Which version should I consume?

Please refer to the table below to:

- Learn if a package is still supported and in which target frameworks.
- Learn what action you should take if you're actively consuming one of these packages:
  - **Remove** the package reference from your projects.
  - **Update** to the latest package version.
  - **Migrate** to another library.
  - *Something else*.

> | Package                                 | Still supported? | Supported where and why?                        | Actions to take                        |

Author: This is the most recent and only feedback item I received:
```suggestion
| Package                                 | Support | Supported where and why?                        | Actions to take                        |
```

| System.Buffers                          | Yes              | <ul><li>.NET Framework 4.6.2+: Provides the implementation of [`ArrayPool<T>`](https://learn.microsoft.com/en-us/dotnet/api/system.buffers.arraypool-1).</li><li>.NET 6+: The type is part of the shared framework.</li></ul> | <ul><li>.NET Framework 4.6.2+: Update.</li><li>.NET 6+: Remove.</li></ul> |
| System.Data.SqlClient                   | Limited          | <ul><li>.NET Framework 4.6.2+: The library is part of the shared framework.</li><li>.NET 8: Marked as obsolete, but [supported until .NET 8 goes EOL.](https://techcommunity.microsoft.com/blog/sqlserver/announcement-system-data-sqlclient-package-is-now-deprecated/4227205)</li><li>All other frameworks: Not supported.</li></ul> | <ul><li> .NET Framework 4.6.2+: Remove.</li><li>.NET 8: System.Data.SqlClient is obsolete. Consider migrating to Microsoft.Data.SqlClient.</li><li>All other frameworks: Remove and migrate to Microsoft.Data.SqlClient.</li></ul> |

> | System.Diagnostics.Tracing              | No               | [`EventSource`](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.tracing.eventsource) is part of the shared framework in all supported frameworks. | Remove. |

Reviewer 2: Why is S.D.Tracing part of this table?

| System.Runtime.WindowsRuntime.UI.Xaml   | No               | Built-in support for WinRT [has been removed from .NET](https://learn.microsoft.com/en-us/dotnet/core/compatibility/interop/5.0/built-in-support-for-winrt-removed).<br />Follow [the recommended actions](https://learn.microsoft.com/en-us/dotnet/core/compatibility/interop/5.0/built-in-support-for-winrt-removed#recommended-action). | Remove. |
| System.Threading.Tasks.Extensions       | Yes              | <ul><li>.NET Framework 4.6.2+: Provides implementations of [`ValueTask`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask) and [`ValueTask<TResult>`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask-1).</li><li>.NET 6+: The types are already part of the shared framework.</li></ul> | <ul><li>.NET Framework 4.6.2+: Update.</li><li>.NET 6+: Remove.</li></ul> |

> | System.ValueTuple                       | No               | The .NET Framework implementation contained in the package is simply an in-box façade that forwards the types to the system-installed framework. | Remove. |

Reviewer 2: Are you sure about this one? I thought System.ValueTuple is only inbox on .NET 4.7+

Author: If I compile this code:
```cs
Console.WriteLine(typeof(System.ValueTuple).Assembly.CodeBase);
```

with this project:
```xml
<Project Sdk="Microsoft.NET.Sdk">
    <PropertyGroup>
        <TargetFrameworks>net462;net47;net472</TargetFrameworks>
        <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
        <OutputType>Exe</OutputType>
    </PropertyGroup>
    <ItemGroup>
      <PackageReference Include="System.ValueTuple" Version="4.5.0" />
    </ItemGroup>
</Project>
```

And then run it against each individual framework (`dotnet run -f <framework>`), I get the same answer for all of them:
```
file:///C:/Windows/Microsoft.NET/Framework64/v4.0.30319/mscorlib.dl
```

In other words, all of them are facades. So the sentence is correct as it is right now.

Reviewer 2: But when I look into the latest package, the `net461` lib assembly has implementation code. Why isn't it used?

Author: I don't know, I didn't check. Do we care? net462 is the minimum supported.

Reviewer 2: The System.ValueTuple package has a `net461` and `net47` asset. Even though .NET Framework 4.6.1 is out-of-support, the package still gets used when you run on .NET Framework 4.6.2. Because net461 is the best matching TFM in the package. So yes, we care.

> And then run it against each individual framework (dotnet run -f <framework>), I get the same answer for all 3:
file:///C:/Windows/Microsoft.NET/Framework64/v4.0.30319/mscorlib.dl

That isn't the right workflow to validate whether package assets are still used. All three .NET Framework TFMs that you target in your app run against the same .NET Framework runtime, presumably 4.8.1. You target different versions but at run time they all use the same underlying run time version. When you run your code against the .NET Framework 4.6.2 run time you will see that the System.ValueTuple OOB package is indeed used.

Author: > That isn't the right workflow to validate whether package assets are still used

That's the method I've been using for all the assemblies we've migrated. `Type.Assembly.CodeBase` tells me during runtime what DLL is being used.

> You target different versions but at run time they all use the same underlying run time version. When you run your code against the .NET Framework 4.6.2

How can I enforce this? @ericstj what are your thoughts on this? I want to make sure I haven't missed any key validations to ensure the migrated assemblies are targeting the right things.

Reviewer 2: You can test this by adding a console log that prints `System.Runtime.InteropServices.RuntimeInformation.FrameworkDescription` (the API unfortunately requires .NET Framework 4.7.1 or above) you will see that the underlying runtime is not 4.6.2.

Reviewer 3: Looking at https://learn.microsoft.com/en-us/dotnet/framework/migration-guide/versions-and-dependencies#net-framework-462 it seems that 4.6.2 was the default in Server 2016, which I suspect is why it's still supported.  I think if you had a VM with this or a windows container you could see that behavior.

Author: Since we have an open issue asking about ValueTuple, we can continue the conversation there: https://github.com/dotnet/maintenance-packages/issues/183

## Reporting issues

If you find an issue in an assembly that is hosted in this repo, please report it in our central repo: https://github.com/dotnet/runtime/issues/new/choose.

## Should I consume this package? Which version should I consume?

Please refer to the table below to:

- Learn if a package is still supported and in which target frameworks.
- Learn what action you should take if you're actively consuming one of these packages.

| Package                                 | Still supported? | Reasons                                         | Actions                                |
|-----------------------------------------|------------------|-------------------------------------------------|----------------------------------------|
| Microsoft.Bcl.HashCode                  | Yes              | Provides the implementation of [`HashCode`](https://learn.microsoft.com/en-us/dotnet/api/system.hashcode) for .NET Framework. | <ul><li>.NET Framework 4.6.2+: Update to latest version of the package.</li><li>.NET 6+: No need to reference the package. The type is part of the shared framework.</li></ul> |

> | Microsoft.CSharp                        | Limited          | Contains an implementation but it is unused in all the currently supported frameworks. | Only reference the package if you're targeting .NET Standard 2.0 and you need to use the Microsoft.CSharp. The implementation found in the package won't be used, but will make sure the namespace is found. The in-box implementation will be used instead. This package won't take any fixes. |

Reviewer 3: Can you somehow make more of the rows use consistent verbiage or style to make it clear how things differ?  Try to limit the actions column to just the framework and the recommendation (remove, update, or avoid use).

```suggestion
| Microsoft.CSharp                        | Limited          | Only used as a reference on .NETStandard 2.0. Other uses are not supported. | <ul><li>.NETStandard 2.0: reference the latest package to use types in the `Microsoft.CSharp` namespace </li><li> All other supported frameworks: remove, the library is part of the framework. |
```

Author: Fixed

| Microsoft.IO.Redist                     | Yes              | Intended for internal use only, not for general use. | Avoid referencing the package. Use the equivalent System.IO APIs instead. |
| System.Buffers                          | Yes              | Provides the implementation of [`ArrayPool<T>`](https://learn.microsoft.com/en-us/dotnet/api/system.buffers.arraypool-1) for .NET Framework. | <ul><li>.NET Framework 4.6.2+: Update to the latest version of the package.</li><li>.NET 6+: No need to reference the package. The types are part of the shared framework.</li></ul> |

> | System.Data.SqlClient                   | Limited          | [Supported until .NET 8 goes EOL.](https://techcommunity.microsoft.com/blog/sqlserver/announcement-system-data-sqlclient-package-is-now-deprecated/4227205) | <ul><li>.NET 4.6.2+ and .NET 6.0+: Either update to the latest version of the package or migrate to Microsoft.Data.SqlClient.</li><li>.NET 9+: System.Data.SqlClient is unsupported. Migrate to Microsoft.Data.SqlClient.</ul> |

Reviewer 4: ```suggestion
| System.Data.SqlClient                   | Limited          | [Supported until .NET 8 goes EOL.](https://techcommunity.microsoft.com/blog/sqlserver/announcement-system-data-sqlclient-package-is-now-deprecated/4227205) | <ul><li> .NET 6.0+ : Either update to the latest version of the package or migrate to Microsoft.Data.SqlClient.</li><li>.NET 9+: System.Data.SqlClient is unsupported. Migrate to Microsoft.Data.SqlClient.</li></ul> |
```

I would not mention framework version, because we don't recommend anyone referencing SqlClient in framework to reference this package.

Reviewer 4: Can we keep it simple for .NET 6.0 and 7.0 and remove the part "Either update to the latest version of the package or migrate to Microsoft.Data.SqlClient."?

> | System.Data.SqlClient                   | Limited          | [Supported until .NET 8 goes EOL.](https://techcommunity.microsoft.com/blog/sqlserver/announcement-system-data-sqlclient-package-is-now-deprecated/4227205) | <ul><li> .NET Framework 4.6.2+: The .NET Framework implementation is simply an in-box façade. No need to reference the package.</li><li> .NET Standard 2.0+, .NET 6.0 and .NET 7.0: System.Data.SqlClient is no longer supported. Migrate to Microsoft.Data.SqlClient.</li><li>.NET 8: System.Data.SqlClient is obsolete. Consider migrating to Microsoft.Data.SqlClient.</li><li>.NET 9+: System.Data.SqlClient is no longer supported. Migrate to Microsoft.Data.SqlClient.</li></ul> |

Reviewer 5: Will users know what "is simply an in-box façade" means? Would "is forwarded to the system-installed .NET Framework" be clearer?

Author: Good point. I rewrote the sentence to include your suggestion.

> | System.Data.SqlClient                   | Limited          | [Supported until .NET 8 goes EOL.](https://techcommunity.microsoft.com/blog/sqlserver/announcement-system-data-sqlclient-package-is-now-deprecated/4227205) | <ul><li> .NET Framework 4.6.2+: The .NET Framework implementation is simply an in-box façade. No need to reference the package.</li><li> .NET Standard 2.0+: No longer supported.</li><li> .NET 6.0 and .NET 7.0: Either update to the latest version of the package or migrate to Microsoft.Data.SqlClient.</li><li>.NET 8: System.Data.SqlClient is obsolete. Consider migrating to Microsoft.Data.SqlClient.</li><li>.NET 9+: System.Data.SqlClient is unsupported. Migrate to Microsoft.Data.SqlClient.</li></ul> |

Author: ```suggestion
| System.Data.SqlClient                   | Limited          | [Supported until .NET 8 goes EOL.](https://techcommunity.microsoft.com/blog/sqlserver/announcement-system-data-sqlclient-package-is-now-deprecated/4227205) | <ul><li> .NET Framework 4.6.2+: The .NET Framework implementation is simply an in-box façade. No need to reference the package.</li><li> .NET Standard 2.0+, .NET 6.0 and .NET 7.0: No longer supported. Migrate to Microsoft.Data.SqlClient.</li><li>.NET 8: System.Data.SqlClient is obsolete. Consider migrating to Microsoft.Data.SqlClient.</li><li>.NET 9+: System.Data.SqlClient is unsupported. Migrate to Microsoft.Data.SqlClient.</li></ul> |
```
